File: Plugins/AM_MSXML_Bruker_PlugIn/clsAnalysisToolRunnerMSXMLBruker.vb
'*********************************************************************************************************
' Written by Matthew Monroe for the US Department of Energy 
' Pacific Northwest National Laboratory, Richland, WA
' Created 03/30/2011
'
'*********************************************************************************************************

Option Strict On

Imports AnalysisManagerBase
Imports System.IO
Imports System.Runtime.InteropServices

Public Class clsAnalysisToolRunnerMSXMLBruker
    Inherits clsAnalysisToolRunnerBase

#Region "Module Variables"
    Protected Const PROGRESS_PCT_MSXML_GEN_RUNNING As Single = 5

    Protected Const COMPASS_XPORT As String = "CompassXport.exe"

    Protected mMSXmlCacheFolder As DirectoryInfo

    Protected WithEvents mCompassXportRunner As clsCompassXportRunner

#End Region

#Region "Methods"

	Protected Const MAX_CSV_FILES As Integer = 50

	''' <summary>
	''' Constructor
	''' </summary>
	''' <remarks>Presently not used</remarks>
	Public Sub New()

	End Sub

	''' <summary>
	''' Runs ReadW tool
	''' </summary>
	''' <returns>CloseOutType enum indicating success or failure</returns>
	''' <remarks></remarks>
	Public Overrides Function RunTool() As IJobParams.CloseOutType
		Dim eResult As IJobParams.CloseOutType
		Dim eReturnCode As IJobParams.CloseOutType

		' Set this to success for now
		eReturnCode = IJobParams.CloseOutType.CLOSEOUT_SUCCESS

		'Do the base class stuff
		If Not MyBase.RunTool = IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
			Return IJobParams.CloseOutType.CLOSEOUT_FAILED
		End If

		' Store the CompassXport version info in the database
		If Not StoreToolVersionInfo() Then
			clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Aborting since StoreToolVersionInfo returned false")
			If String.IsNullOrEmpty(m_message) Then
				m_message = "Error determining CompassXport version"
			End If
			Return IJobParams.CloseOutType.CLOSEOUT_FAILED
		End If

        Dim msXMLCacheFolderPath As String = m_mgrParams.GetParam("MSXMLCacheFolderPath", String.Empty)
        mMSXmlCacheFolder = New DirectoryInfo(msXMLCacheFolderPath)

        If Not mMSXmlCacheFolder.Exists Then
            LogError("MSXmlCache folder not found: " & msXMLCacheFolderPath)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

		Dim processingErrorMessage As String = String.empty
        Dim fiResultsFile As FileInfo = Nothing

        eResult = CreateMSXmlFile(fiResultsFile)

		If eResult <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
			' Something went wrong
			' In order to help diagnose things, we will move whatever files were created into the eResult folder, 
			'  archive it using CopyFailedResultsToArchiveFolder, then return IJobParams.CloseOutType.CLOSEOUT_FAILED
			If String.IsNullOrEmpty(m_message) Then
				m_message = "Error running CompassXport"
			End If
			processingErrorMessage = String.copy(m_message)

			If eResult = IJobParams.CloseOutType.CLOSEOUT_NO_DATA Then
				eReturnCode = eResult
			Else
				eReturnCode = IJobParams.CloseOutType.CLOSEOUT_FAILED
			End If
        Else
            ' Gzip the .mzML or .mzXML file then copy to the server cache
            eReturnCode = PostProcessMsXmlFile(fiResultsFile)
        End If

        'Stop the job timer
        m_StopTime = DateTime.UtcNow

        'Delete the raw data files
        If m_DebugLevel > 3 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsAnalysisToolRunnerMSXMLBruker.RunTool(), Deleting raw data file")
        End If

        If DeleteRawDataFiles() <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "clsAnalysisToolRunnerMSXMLBruker.RunTool(), Problem deleting raw data files: " & m_message)

            If Not String.IsNullOrEmpty(processingErrorMessage) Then
                m_message = processingErrorMessage
            Else
                ' Don't treat this as a critical error; leave eReturnCode unchanged
                m_message = "Error deleting raw data files"
            End If
        End If

        'Update the job summary file
        If m_DebugLevel > 3 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsAnalysisToolRunnerMSXMLBruker.RunTool(), Updating summary file")
        End If
        UpdateSummaryFile()

        'Make the results folder
        If m_DebugLevel > 3 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsAnalysisToolRunnerMSXMLBruker.RunTool(), Making results folder")
        End If

        eResult = MakeResultsFolder()
        If eResult <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            ' MakeResultsFolder handles posting to local log, so set database error message and exit
            m_message = "Error making results folder"
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        eResult = MoveResultFiles()
        If eResult <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            ' MoveResultFiles moves the eResult files to the eResult folder
            m_message = "Error moving files into results folder"
            eReturnCode = IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        If eReturnCode = IJobParams.CloseOutType.CLOSEOUT_FAILED Then
            ' Try to save whatever files were moved into the results folder
            Dim objAnalysisResults = New clsAnalysisResults(m_mgrParams, m_jobParams)
            objAnalysisResults.CopyFailedResultsToArchiveFolder(Path.Combine(m_WorkDir, m_ResFolderName))

            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        eResult = CopyResultsFolderToServer()
        If eResult <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
            Return eResult
        End If

        'If we get to here, everything worked so exit happily
        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

    End Function

    ''' <summary>
    ''' Generate the mzXML or mzML file
    ''' </summary>
    ''' <returns>CloseOutType enum indicating success or failure</returns>
    ''' <remarks></remarks>
    Private Function CreateMSXmlFile(<Out()> ByRef fiResultsFile As FileInfo) As IJobParams.CloseOutType

        If m_DebugLevel > 4 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "clsAnalysisToolRunnerMSXMLGen.CreateMSXmlFile(): Enter")
        End If

        Dim msXmlGenerator As String = m_jobParams.GetParam("MSXMLGenerator")           ' Typically CompassXport.exe

        Dim msXmlFormat As String = m_jobParams.GetParam("MSXMLOutputType")             ' Typically mzXML or mzML
        Dim CentroidMSXML = CBool(m_jobParams.GetParam("CentroidMSXML"))

        Dim CompassXportProgramPath As String
        Dim eOutputType As clsCompassXportRunner.MSXMLOutputTypeConstants

        Dim blnSuccess As Boolean

        ' Initialize the Results File output parameter to a dummy name for now
        fiResultsFile = New FileInfo(Path.Combine(m_WorkDir, "NonExistent_Placeholder_File.tmp"))

        If String.Equals(msXmlGenerator, COMPASS_XPORT, StringComparison.CurrentCultureIgnoreCase) Then
            CompassXportProgramPath = m_mgrParams.GetParam("CompassXportLoc")

            If String.IsNullOrEmpty(CompassXportProgramPath) Then
                m_message = "Manager parameter CompassXportLoc is not defined in the Manager Control DB"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            If Not File.Exists(CompassXportProgramPath) Then
                m_message = "CompassXport program not found"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & " at " & CompassXportProgramPath)
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If
        Else
            m_message = "Invalid value for MSXMLGenerator: " & msXmlGenerator
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        eOutputType = clsCompassXportRunner.GetMsXmlOutputTypeByName(msXmlFormat)
        If eOutputType = clsCompassXportRunner.MSXMLOutputTypeConstants.Invalid Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "msXmlFormat string is not recognized (" & msXmlFormat & "); it is typically mzXML, mzML, or CSV; will default to mzXML")
            eOutputType = clsCompassXportRunner.MSXMLOutputTypeConstants.mzXML
        End If

        fiResultsFile = New FileInfo(Path.Combine(m_WorkDir, m_Dataset & "." & clsCompassXportRunner.GetMsXmlOutputTypeByID(eOutputType)))

        ' Instantiate the processing class
        mCompassXportRunner = New clsCompassXportRunner(m_WorkDir, CompassXportProgramPath, m_Dataset, eOutputType, CentroidMSXML)

        ' Create the file
        blnSuccess = mCompassXportRunner.CreateMSXMLFile

        If Not blnSuccess Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, mCompassXportRunner.ErrorMessage)
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED

        ElseIf mCompassXportRunner.ErrorMessage.Length > 0 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, mCompassXportRunner.ErrorMessage)

        End If

        If eOutputType <> clsCompassXportRunner.MSXMLOutputTypeConstants.CSV Then
            If fiResultsFile.Exists Then
                Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS
            End If

            m_message = "MSXml results file not found: " & fiResultsFile.FullName
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
            Return IJobParams.CloseOutType.CLOSEOUT_FILE_NOT_FOUND
        End If

        ' CompassXport created one CSV file for each spectrum in the dataset
        ' Confirm that fewer than 100 CSV files were created

        Dim diWorkDir As New DirectoryInfo(m_WorkDir)
        Dim fiFiles As List(Of FileInfo) = diWorkDir.GetFiles("*.csv").ToList()

        If fiFiles.Count < MAX_CSV_FILES Then
            Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS
        End If

        m_message = "CompassXport created " & fiFiles.Count.ToString() & " CSV files. The CSV conversion mode is only appropriate for datasets with fewer than " & MAX_CSV_FILES & " spectra; create a mzXML file instead (e.g., settings file mzXML_Bruker.xml)"
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)

        For Each fiFile In fiFiles
            Try
                fiFile.Delete()
            Catch ex As Exception
                ' Ignore errors here
            End Try
        Next

        Return IJobParams.CloseOutType.CLOSEOUT_FAILED

    End Function

    Private Function PostProcessMsXmlFile(fiResultsFile As FileInfo) As IJobParams.CloseOutType

        ' Compress the file using GZip
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, "GZipping " & fiResultsFile.Name)
        fiResultsFile = GZipFile(fiResultsFile)

        If fiResultsFile Is Nothing Then
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        ' Copy the .mzXML.gz or .mzML.gz file to the MSXML cache
        Dim remoteCachefilePath = CopyFileToServerCache(mMSXmlCacheFolder.FullName, fiResultsFile.FullName, purgeOldFilesIfNeeded:=True)

        If String.IsNullOrEmpty(remoteCachefilePath) Then
            If String.IsNullOrEmpty(m_message) Then
                LogError("CopyFileToServerCache returned false for " & fiResultsFile.Name)
            End If
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End If

        ' Create the _CacheInfo.txt file
        Dim cacheInfoFilePath = fiResultsFile.FullName & "_CacheInfo.txt"
        Using swOutFile = New StreamWriter(New FileStream(cacheInfoFilePath, FileMode.Create, FileAccess.Write, FileShare.Read))
            swOutFile.WriteLine(remoteCachefilePath)
        End Using

        m_jobParams.AddResultFileToSkip(fiResultsFile.Name)

        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

    End Function

    ''' <summary>
    ''' Stores the tool version info in the database
    ''' </summary>
    ''' <remarks></remarks>
    Protected Function StoreToolVersionInfo() As Boolean

        Dim strToolVersionInfo As String = String.Empty

        If m_DebugLevel >= 2 Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, "Determining tool version info")
        End If

        ' Store paths to key files in ioToolFiles
        Dim ioToolFiles As New List(Of FileInfo)

        Dim msXmlGenerator As String = m_jobParams.GetParam("MSXMLGenerator")           ' Typically CompassXport.exe
        If String.Equals(msXmlGenerator, COMPASS_XPORT, StringComparison.CurrentCultureIgnoreCase) Then
            Dim compassXportPath = m_mgrParams.GetParam("CompassXportLoc")
            If String.IsNullOrEmpty(compassXportPath) Then
                m_message = "Path defined by manager param CompassXportLoc is empty"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
                Return False
            End If

            Try
                ioToolFiles.Add(New FileInfo(compassXportPath))
            Catch ex As Exception
                m_message = "Path defined by manager param CompassXportLoc is invalid: " + compassXportPath
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message + "; " + ex.Message)
                Return False
            End Try

        Else
            If String.IsNullOrEmpty(msXmlGenerator) Then
                m_message = "Job Parameter MSXMLGenerator is not defined"
            Else
                m_message = "Invalid value for MSXMLGenerator, should be " & COMPASS_XPORT & ", not " & msXmlGenerator
            End If

            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
            Return False
        End If

        Try
            Return MyBase.SetStepTaskToolVersion(strToolVersionInfo, ioToolFiles)
        Catch ex As Exception
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, "Exception calling SetStepTaskToolVersion: " & ex.Message)
            Return False
        End Try

    End Function

#End Region

#Region "Event Handlers"
    ''' <summary>
    ''' Event handler for CompassXportRunner.LoopWaiting event
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub CompassXportRunner_LoopWaiting() Handles mCompassXportRunner.LoopWaiting

        UpdateStatusFile(PROGRESS_PCT_MSXML_GEN_RUNNING)

        LogProgress("CompassXport")

    End Sub

    ''' <summary>
    ''' Event handler for mCompassXportRunner.ProgRunnerStarting event
    ''' </summary>
    ''' <param name="CommandLine">The command being executed (program path plus command line arguments)</param>
    ''' <remarks></remarks>
    Private Sub mCompassXportRunner_ProgRunnerStarting(CommandLine As String) Handles mCompassXportRunner.ProgRunnerStarting
        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.DEBUG, CommandLine)
    End Sub
#End Region

End Class

File: Plugins/AM_MSXML_Bruker_PlugIn/clsAnalysisToolRunnerMSXMLBruker.cs
﻿//*********************************************************************************************************
// Written by Matthew Monroe for the US Department of Energy
// Pacific Northwest National Laboratory, Richland, WA
// Created 03/30/2011
//
//*********************************************************************************************************

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AnalysisManagerBase;

namespace AnalysisManagerMsXmlBrukerPlugIn
{

    /// <summary>
    /// Class for running MSXML Bruker
    /// </summary>
    public class clsAnalysisToolRunnerMSXMLBruker : clsAnalysisToolRunnerBase
    {
        #region "Module Variables"

        protected const float PROGRESS_PCT_MSXML_GEN_RUNNING = 5;

        protected const string COMPASS_XPORT = "CompassXport.exe";

        protected DirectoryInfo mMSXmlCacheFolder;

        protected clsCompassXportRunner mCompassXportRunner;

        #endregion

        #region "Methods"

        protected const int MAX_CSV_FILES = 50;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <remarks>Presently not used</remarks>
        // ReSharper disable once EmptyConstructor
        public clsAnalysisToolRunnerMSXMLBruker()
        {
            // Empty constructor
        }

        /// <summary>
        /// Runs ReadW tool
        /// </summary>
        /// <returns>CloseOutType enum indicating success or failure</returns>
        public override CloseOutType RunTool()
        {
            // Do the base class stuff
            if (base.RunTool() != CloseOutType.CLOSEOUT_SUCCESS)
            {
                return CloseOutType.CLOSEOUT_FAILED;
            }

            // Store the CompassXport version info in the database
            if (!StoreToolVersionInfo())
            {
                LogError("Aborting since StoreToolVersionInfo returned false");
                if (string.IsNullOrEmpty(mMessage))
                {
                    mMessage = "Error determining CompassXport version";
                }
                return CloseOutType.CLOSEOUT_FAILED;
            }

            var msXMLCacheFolderPath = mMgrParams.GetParam("MSXMLCacheFolderPath", string.Empty);
            mMSXmlCacheFolder = new DirectoryInfo(msXMLCacheFolderPath);

            if (!mMSXmlCacheFolder.Exists)
            {
                LogError("MSXmlCache folder not found: " + msXMLCacheFolderPath);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            var processingErrorMessage = string.Empty;

            var eResult = CreateMSXmlFile(out var resultsFile);

            if (eResult != CloseOutType.CLOSEOUT_SUCCESS)
            {
                // Something went wrong
                // In order to help diagnose things, we will move whatever files were created into the eResult folder,
                //  archive it using CopyFailedResultsToArchiveDirectory, then return CloseOutType.CLOSEOUT_FAILED
                if (string.IsNullOrEmpty(mMessage))
                {
                    mMessage = "Error running CompassXport";
                }
                processingErrorMessage = string.Copy(mMessage);
            }
            else
            {
                // Gzip the .mzML or .mzXML file then copy to the server cache
                PostProcessMsXmlFile(resultsFile);
            }

            // Stop the job timer
            mStopTime = DateTime.UtcNow;

            // Delete the raw data files
            if (mDebugLevel > 3)
            {
                LogDebug("clsAnalysisToolRunnerMSXMLBruker.RunTool(), Deleting raw data file");
            }

            var deleteSuccess = DeleteRawDataFiles();
            if (!deleteSuccess)
            {
                LogError("clsAnalysisToolRunnerMSXMLBruker.RunTool(), Problem deleting raw data files: " + mMessage);

                if (!string.IsNullOrEmpty(processingErrorMessage))
                {
                    mMessage = processingErrorMessage;
                }
                else
                {
                    // Don't treat this as a critical error; leave eReturnCode unchanged
                    mMessage = "Error deleting raw data files";
                }
            }

            // Update the job summary file
            if (mDebugLevel > 3)
            {
                LogDebug("clsAnalysisToolRunnerMSXMLBruker.RunTool(), Updating summary file");
            }

            UpdateSummaryFile();

            var success = CopyResultsToTransferDirectory();

            if (success && eResult == CloseOutType.CLOSEOUT_SUCCESS)
                return CloseOutType.CLOSEOUT_SUCCESS;

            return CloseOutType.CLOSEOUT_FAILED;

        }

        /// <summary>
        /// Generate the mzXML or mzML file
        /// </summary>
        /// <returns>CloseOutType enum indicating success or failure</returns>
        /// <remarks></remarks>
        private CloseOutType CreateMSXmlFile(out FileInfo resultsFile)
        {
            if (mDebugLevel > 4)
            {
                LogDebug("clsAnalysisToolRunnerMSXMLGen.CreateMSXmlFile(): Enter");
            }

            var msXmlGenerator = mJobParams.GetParam("MSXMLGenerator");
            // Typically CompassXport.exe

            var msXmlFormat = mJobParams.GetParam("MSXMLOutputType");
            // Typically mzXML or mzML
            var CentroidMSXML = Convert.ToBoolean(mJobParams.GetParam("CentroidMSXML"));

            string CompassXportProgramPath;

            // Initialize the Results File output parameter to a dummy name for now
            resultsFile = new FileInfo(Path.Combine(mWorkDir, "NonExistent_Placeholder_File.tmp"));

            if (string.Equals(msXmlGenerator, COMPASS_XPORT, StringComparison.OrdinalIgnoreCase))
            {
                CompassXportProgramPath = mMgrParams.GetParam("CompassXportLoc");

                if (string.IsNullOrEmpty(CompassXportProgramPath))
                {
                    mMessage = "Manager parameter CompassXportLoc is not defined in the Manager Control DB";
                    LogError(mMessage);
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                if (!File.Exists(CompassXportProgramPath))
                {
                    mMessage = "CompassXport program not found";
                    LogError(mMessage + " at " + CompassXportProgramPath);
                    return CloseOutType.CLOSEOUT_FAILED;
                }
            }
            else
            {
                mMessage = "Invalid value for MSXMLGenerator: " + msXmlGenerator;
                LogError(mMessage);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            var eOutputType = clsCompassXportRunner.GetMsXmlOutputTypeByName(msXmlFormat);
            if (eOutputType == clsCompassXportRunner.MSXMLOutputTypeConstants.Invalid)
            {
                LogWarning("msXmlFormat string is not recognized (" + msXmlFormat + "); it is typically mzXML, mzML, or CSV; will default to mzXML");
                eOutputType = clsCompassXportRunner.MSXMLOutputTypeConstants.mzXML;
            }

            resultsFile = new FileInfo(Path.Combine(mWorkDir, mDatasetName + "." + clsCompassXportRunner.GetMsXmlOutputTypeByID(eOutputType)));

            // Instantiate the processing class
            mCompassXportRunner = new clsCompassXportRunner(mWorkDir, CompassXportProgramPath, mDatasetName, eOutputType, CentroidMSXML);
            RegisterEvents(mCompassXportRunner);

            mCompassXportRunner.LoopWaiting += CompassXportRunner_LoopWaiting;
            mCompassXportRunner.ProgRunnerStarting += CompassXportRunner_ProgRunnerStarting;

            // Create the file
            var success = mCompassXportRunner.CreateMSXMLFile();

            if (!success)
            {
                LogWarning(mCompassXportRunner.ErrorMessage);
                return CloseOutType.CLOSEOUT_FAILED;
            }

            if (mCompassXportRunner.ErrorMessage.Length > 0)
            {
                LogWarning(mCompassXportRunner.ErrorMessage);
            }

            if (eOutputType != clsCompassXportRunner.MSXMLOutputTypeConstants.CSV)
            {
                if (resultsFile.Exists)
                {
                    return CloseOutType.CLOSEOUT_SUCCESS;
                }

                mMessage = "MSXml results file not found: " + resultsFile.FullName;
                LogError(mMessage);
                return CloseOutType.CLOSEOUT_FILE_NOT_FOUND;
            }

            // CompassXport created one CSV file for each spectrum in the dataset
            // Confirm that fewer than 100 CSV files were created

            var workDir = new DirectoryInfo(mWorkDir);
            var csvFiles = workDir.GetFiles("*.csv").ToList();

            if (csvFiles.Count < MAX_CSV_FILES)
            {
                return CloseOutType.CLOSEOUT_SUCCESS;
            }

            mMessage = "CompassXport created " + csvFiles.Count +
                        " CSV files. The CSV conversion mode is only appropriate for datasets with fewer than " + MAX_CSV_FILES +
                        " spectra; create a mzXML file instead (e.g., settings file mzXML_Bruker.xml)";
            LogError(mMessage);

            foreach (var csvFile in csvFiles)
            {
                try
                {
                    csvFile.Delete();
                }
                catch (Exception)
                {
                    // Ignore errors here
                }
            }

            return CloseOutType.CLOSEOUT_FAILED;
        }

        private void PostProcessMsXmlFile(FileInfo resultsFile)
        {
            // Compress the file using GZip
            LogMessage("GZipping " + resultsFile.Name);
            resultsFile = GZipFile(resultsFile);

            if (resultsFile == null)
            {
                return;
            }

            // Copy the .mzXML.gz or .mzML.gz file to the MSXML cache
            var remoteCachefilePath = CopyFileToServerCache(mMSXmlCacheFolder.FullName, resultsFile.FullName, purgeOldFilesIfNeeded: true);

            if (string.IsNullOrEmpty(remoteCachefilePath))
            {
                if (string.IsNullOrEmpty(mMessage))
                {
                    LogError("CopyFileToServerCache returned false for " + resultsFile.Name);
                }

                return;
            }

            // Create the _CacheInfo.txt file
            var cacheInfoFilePath = resultsFile.FullName + "_CacheInfo.txt";
            using (var writer = new StreamWriter(new FileStream(cacheInfoFilePath, FileMode.Create, FileAccess.Write, FileShare.Read)))
            {
                writer.WriteLine(remoteCachefilePath);
            }

            mJobParams.AddResultFileToSkip(resultsFile.Name);

        }

        /// <summary>
        /// Stores the tool version info in the database
        /// </summary>
        /// <remarks></remarks>
        protected bool StoreToolVersionInfo()
        {
            var toolVersionInfo = string.Empty;

            if (mDebugLevel >= 2)
            {
                LogDebug("Determining tool version info");
            }

            // Store paths to key files in toolFiles
            var toolFiles = new List<FileInfo>();

            var msXmlGenerator = mJobParams.GetParam("MSXMLGenerator");

            // Typically CompassXport.exe
            if (string.Equals(msXmlGenerator, COMPASS_XPORT, StringComparison.OrdinalIgnoreCase))
            {
                var compassXportPath = mMgrParams.GetParam("CompassXportLoc");
                if (string.IsNullOrEmpty(compassXportPath))
                {
                    mMessage = "Path defined by manager param CompassXportLoc is empty";
                    LogError(mMessage);
                    return false;
                }

                try
                {
                    toolFiles.Add(new FileInfo(compassXportPath));
                }
                catch (Exception ex)
                {
                    mMessage = "Path defined by manager param CompassXportLoc is invalid: " + compassXportPath;
                    LogError(mMessage + "; " + ex.Message);
                    return false;
                }
            }
            else
            {
                if (string.IsNullOrEmpty(msXmlGenerator))
                {
                    mMessage = "Job Parameter MSXMLGenerator is not defined";
                }
                else
                {
                    mMessage = "Invalid value for MSXMLGenerator, should be " + COMPASS_XPORT + ", not " + msXmlGenerator;
                }

                LogError(mMessage);
                return false;
            }

            try
            {
                return SetStepTaskToolVersion(toolVersionInfo, toolFiles);
            }
            catch (Exception ex)
            {
                LogError("Exception calling SetStepTaskToolVersion: " + ex.Message);
                return false;
            }
        }

        #endregion

        #region "Event Handlers"

        /// <summary>
        /// Event handler for CompassXportRunner.LoopWaiting event
        /// </summary>
        /// <remarks></remarks>
        private void CompassXportRunner_LoopWaiting()
        {
            UpdateStatusFile(PROGRESS_PCT_MSXML_GEN_RUNNING);

            LogProgress("CompassXport");
        }

        /// <summary>
        /// Event handler for mCompassXportRunner.ProgRunnerStarting event
        /// </summary>
        /// <param name="commandLine">The command being executed (program path plus command line arguments)</param>
        /// <remarks></remarks>
        private void CompassXportRunner_ProgRunnerStarting(string commandLine)
        {
            LogDebug(commandLine);
        }

        #endregion
    }
}
